src/main/kotlin/no/fdk/userapi/adapter/AltinnAdapter.kt

```diff
@@ -1,63 +1,91 @@
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import io.netty.handler.timeout.ReadTimeoutHandler
+import io.netty.handler.timeout.WriteTimeoutHandler
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.reactor.awaitSingle
+import kotlinx.coroutines.withContext
 import no.fdk.userapi.configuration.HostProperties
 import no.fdk.userapi.mapper.toOrganization
 import no.fdk.userapi.mapper.toPerson
 import no.fdk.userapi.model.*
 import org.slf4j.LoggerFactory
-import org.springframework.http.HttpStatus
+import org.springframework.cache.CacheManager
+import org.springframework.cache.get
+import org.springframework.http.client.reactive.ReactorClientHttpConnector
 import org.springframework.stereotype.Service
-import java.io.InputStream
-import java.net.HttpURLConnection
+import org.springframework.web.reactive.function.client.WebClient
+import reactor.netty.http.client.HttpClient
+import reactor.netty.resources.ConnectionProvider
 import java.net.URI
+import java.time.Duration
 
 private val logger = LoggerFactory.getLogger(AltinnAdapter::class.java)
 
 @Service
-class AltinnAdapter(private val hostProperties: HostProperties) {
-    private val fiveSeconds = 5000
-    private val twentySeconds = 20000
+class AltinnAdapter(private val hostProperties: HostProperties, private val cacheManager: CacheManager) {
+    private val webClient = WebClient.builder()
+        .clientConnector(ReactorClientHttpConnector(
+            HttpClient.create(
+                ConnectionProvider.builder("custom")
+                    .maxConnections(100)
+                    .pendingAcquireTimeout(Duration.ofSeconds(60))
+                    .build())
+            .responseTimeout(Duration.ofSeconds(30))
+            .doOnConnected { conn ->
+                conn.addHandlerLast(ReadTimeoutHandler(30))
+                    .addHandlerLast(WriteTimeoutHandler(30))
+            }))
+        .build()
 
-    private fun altinnStream(uri: URI): InputStream =
-        with(uri.toURL().openConnection() as HttpURLConnection) {
-            connectTimeout = fiveSeconds
-            readTimeout = twentySeconds
-            connect()
-            if (HttpStatus.resolve(responseCode)?.is2xxSuccessful == true) {
-                inputStream
-            } else {
-                throw Exception("altinn connection failed with code $responseCode")
-            }
+    private suspend fun <T: Any> fetchJson(uri: URI, typeRef: TypeReference<T>): T? = withContext(Dispatchers.IO) {
+        val cache = cacheManager["altinn"]
+        return@withContext cache?.get(uri.toString())?.let { it.get() as T? } ?: run {
+            logger.debug("Fetching JSON")
+            val response = webClient.get()
+                .uri(uri)
+                .retrieve()
+                .bodyToMono(String::class.java)
+                .awaitSingle()
+
+            val jsonObject = ObjectMapper().readValue(response, typeRef)
```

Comment:
Bruk den prekonfigurerte kotlin-varianten
```suggestion
            val jsonObject = jacksonObjectMapper().readValue(response, typeRef)
```

Reply:
Eller registrer denne med den modifiserbare modulen
```suggestion
            val jsonObject = ObjectMapper().registerModule(KotlinModule.Builder().build()).readValue(response, typeRef)
```